x/storage/keeper/keeper.go

```diff
@@ -1464,3 +1479,140 @@ func (k Keeper) getDiscontinueObjectStatus(ctx sdk.Context, objectId types.Uint)
+
+	// persist current block stale permission info to store if exists
+	if !deleteInfo.IsEmpty() {
+		store := ctx.KVStore(k.storeKey)
+		store.Set(types.GetDeleteStalePoliciesKey(ctx.BlockHeight()), bz)
+	}
+}
+
+func (k Keeper) GarbageCollectResourcesStalePolicy(ctx sdk.Context) {
+	store := ctx.KVStore(k.storeKey)
+	deleteStalePoliciesPrefixStore := prefix.NewStore(store, types.DeleteStalePoliciesPrefix)
+
+	// todo use store to keep track of processed GC height
+	iterator := deleteStalePoliciesPrefixStore.Iterator(nil, nil)
+	defer iterator.Close()
+
+	maxCleanup := k.StalePolicyCleanupMax(ctx)
+	deletedCount := uint64(0)
+
+	for ; iterator.Valid(); iterator.Next() {
+		deleteInfo := &types.DeleteInfo{}
+		k.cdc.MustUnmarshal(iterator.Value(), deleteInfo)
+
+		if deleteInfo.ObjectIds != nil && len(deleteInfo.ObjectIds.Id) > 0 {
+			ids := deleteInfo.ObjectIds.Id
+			for idx, id := range ids {
+				k.permKeeper.ForceDeleteAccountPolicyForResource(ctx, resource.RESOURCE_TYPE_OBJECT, id)
+				k.permKeeper.ForceDeleteGroupPolicyForResource(ctx, resource.RESOURCE_TYPE_OBJECT, id)
+				deletedCount++
+				// reaches the deletion limit during current endblocker
+				if deletedCount > maxCleanup {
+					ids = ids[idx+1:]
+					deleteInfo.ObjectIds.Id = ids
+					deleteStalePoliciesPrefixStore.Set(iterator.Key(), k.cdc.MustMarshal(deleteInfo))
+					return
+				}
+			}
+			// clean deleted objects id from deleteInfo
+			deleteInfo.ObjectIds = nil
+		}
+
+		if deleteInfo.BucketIds != nil && len(deleteInfo.BucketIds.Id) > 0 {
+			ids := deleteInfo.BucketIds.Id
+			for idx, id := range ids {
+				k.permKeeper.ForceDeleteAccountPolicyForResource(ctx, resource.RESOURCE_TYPE_BUCKET, id)
+				k.permKeeper.ForceDeleteGroupPolicyForResource(ctx, resource.RESOURCE_TYPE_BUCKET, id)
+				deletedCount++
+				if deletedCount > maxCleanup {
+					ids = ids[idx+1:]
+					deleteInfo.BucketIds.Id = ids
+					deleteStalePoliciesPrefixStore.Set(iterator.Key(), k.cdc.MustMarshal(deleteInfo))
+					return
+				}
+			}
+			// clean deleted buckets id from deleteInfo
+			deleteInfo.BucketIds = nil
+		}
+
+		if deleteInfo.GroupIds != nil && len(deleteInfo.GroupIds.Id) > 0 {
+			ids := deleteInfo.GroupIds.Id
+			// for a group which is deleted by user, there are 2 parts need to clean:
+			// 1. group members within the group.
+			// 2. group policy
+			for idx, id := range ids {
+				k.permKeeper.ForceDeleteGroupMembers(ctx, id)
+				k.permKeeper.ForceDeleteAccountPolicyForResource(ctx, resource.RESOURCE_TYPE_GROUP, id)
+				deletedCount++
+				if deletedCount > maxCleanup {
+					ids = ids[idx+1:]
+					deleteInfo.GroupIds.Id = ids
+					deleteStalePoliciesPrefixStore.Set(iterator.Key(), k.cdc.MustMarshal(deleteInfo))
+					return
+				}
+			}
+			// clean deleted buckets id from deleteInfo
+			deleteInfo.GroupIds = nil
+		}
+
+		// the specified block height(iterator-key)'s stale resource permission metadata is purged
+		if deleteInfo.IsEmpty() {
```

Comment:
If the delete info is not empty, but some permission info associated with certain resources has already been deleted, it still needs persist.